File: Python/Linear optimization lesson/V2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<double> get_vector(int size)
{
    vector<double> v(size);
    for (double &val : v)
    {
        cin >> val;
    }
    return v;
}

vector<vector<double>> get_matrix(int rows, int cols)
{
    vector<vector<double>> m(rows, vector<double>(cols));
    for (vector<double> &row : m)
    {
        for (double &val : row)
        {
            cin >> val;
        }
    }
    return m;
}

void print_tableau(vector<vector<double>> &tableau)
{
    for (auto &row : tableau)
    {
        for (auto &val : row)
        {
            cout << val << ' ';
        }
        cout << '\n';
    }
    cout << '\n';
}

vector<vector<double>> revised_simplex(vector<double> c, vector<vector<double>> A, vector<double> b)
{
    int num_vars = c.size();
    int num_constraints = b.size();

    // Initialize the tableau
    vector<vector<double>> tableau(num_constraints + 1, vector<double>(num_vars + num_constraints + 1));
    for (int i = 0; i < num_constraints; ++i)
    {
        for (int j = 0; j < num_vars; ++j)
        {
            tableau[i][j] = A[i][j];
        }
        tableau[i][num_vars + i] = 1;
        tableau[i][num_vars + num_constraints] = b[i];
    }
    for (int j = 0; j < num_vars; ++j)
    {
        tableau[num_constraints][j] = c[j];
    }

    print_tableau(tableau);

    // Main loop
    while (true)
    {
        int pivot_column = max_element(tableau[num_constraints].begin(), tableau[num_constraints].end()) - tableau[num_constraints].begin();
        if (tableau[num_constraints][pivot_column] <= 0)
            break;

        pair<double, int> min_ratio = {1e9, -1};
        for (int i = 0; i < num_constraints; ++i)
        {
            if (tableau[i][pivot_column] > 0 && tableau[i][num_vars + num_constraints] / tableau[i][pivot_column] < min_ratio.first)
            {
                min_ratio = {tableau[i][num_vars + num_constraints] / tableau[i][pivot_column], i};
            }
        }
        int pivot_row = min_ratio.second;

        double pivot_element = tableau[pivot_row][pivot_column];
        for (int j = 0; j <= num_vars + num_constraints; ++j)
        {
            tableau[pivot_row][j] /= pivot_element;
        }
        for (int i = 0; i <= num_constraints; ++i)
        {
            if (i != pivot_row)
            {
                double multiplier = tableau[i][pivot_column];
                for (int j = 0; j <= num_vars + num_constraints; ++j)
                {
                    tableau[i][j] -= multiplier * tableau[pivot_row][j];
                }
            }
        }

        print_tableau(tableau);
    }

    return tableau;
}

int main()
{
    int num_vars, num_constraints;
    cout << "Enter the number of variables: ";
    cin >> num_vars;
    cout << "Enter the number of constraints: ";
    cin >> num_constraints;

    cout << "Enter the coefficients of the objective function: ";
    vector<double> c = get_vector(num_vars);

    cout << "Enter the coefficients of the constraints: ";
    vector<vector<double>> A = get_matrix(num_constraints, num_vars);

    cout << "Enter the RHS values of the constraints: ";
    vector<double> b = get_vector(num_constraints);

    vector<vector<double>> tableau = revised_simplex(c, A, b);
    cout << "Optimal solution: " << tableau.back().back() << '\n';
    return 0;
}

File: Python/Linear optimization lesson/V2.py
import numpy as np


def get_input():
    c = list(map(float, input(
        "Enter the coefficients of the objective function (separated by spaces): ").split()))
    num_constraints = int(input("Enter the number of constraints: "))
    A = []
    for i in range(num_constraints):
        A.append(list(map(float, input(
            f"Enter the coefficients of constraint {i+1} (separated by spaces): ").split())))
    b = list(map(float, input(
        "Enter the RHS values of the constraints (separated by spaces): ").split()))
    return c, A, b


def print_tableau(tableau, step):
    print(f"Step {step}:")
    print(tableau)
    print()


def revised_simplex(c, A, b):
    num_vars = len(c)
    num_constraints = len(b)

    # Initialize the tableau
    tableau = np.zeros((num_constraints + 1, num_vars + num_constraints + 1))
    tableau[:-1, :num_vars] = A
    tableau[:-1, num_vars:num_vars+num_constraints] = np.eye(num_constraints)
    tableau[:-1, -1] = b
    tableau[-1, :num_vars] = c
    tableau[-1, -1] = 0

    step = 0
    print_tableau(tableau, step)

    # Main loop
    while np.any(tableau[-1, :-1] > 0):
        pivot_column = np.argmax(tableau[-1, :-1])
        ratios = tableau[:-1, -1] / tableau[:-1, pivot_column]
        pivot_row = np.argmin(np.where(ratios > 0, ratios, np.inf))
        pivot_element = tableau[pivot_row, pivot_column]
        tableau[pivot_row, :] /= pivot_element
        for i in range(len(tableau)):
            if i != pivot_row:
                tableau[i, :] -= tableau[i, pivot_column] * \
                    tableau[pivot_row, :]
        step += 1
        print_tableau(tableau, step)
    return tableau


# Example usage:
c, A, b = get_input()
tableau = revised_simplex(c, A, b)
print("Optimal solution:", tableau[-1, -1])
